OPEN_PROBLEMS/ENHANCED_BITSWAP_GRAPHSYNC.md

```diff
@@ -0,0 +1,128 @@
+
+## Short Description
+> In one sentence or paragraph.
+
+Once IPFS has resolved the CID to a peerID and further to the exact location of the node that  stores the requested content, it then uses a very simple protocol called bitswap to exchange content between the requestor and the server.
+
+Although bitswap is simple and generally works, *its performance is suboptimal*. This is mainly due to the fact that a node cannot request a subgraph of the DAG and results in many round-trips in order to “walk down” the DAG. The current operation of bitswap is also very often linked to duplicate transmission and receipt of content which overloads both the end nodes and the network.
+
+## Long Description
+
+In order to synchronise a DAG between untrusted nodes, bitswap is exploiting the content-addressability feature of IPFS. This means that bitswap starts with the root hash of the DAG. Once it fetches the data, it verifies that its hash matches. The node can now trust this block and can thus continue with the rest of the blocks in the DAG (aka walk the DAG) until it gets the complete DAG and the data associated with it.
+
+In the current implementation of bitswap, requesting nodes send their WANT lists to all the peers they are directly connected to. This results in potentially duplicate data being sent back to the receiving node. When the receiving node has received the block(s) it asked for, it sends a CANCEL message to its peers to let them know that the block(s) is not in its WANT list anymore.
+
+If none of the directly connected peers have any of the WANT list blocks, bitswap falls back to the DHT to find the requested content. This results in long delays to get to a peer that stores the requested content.
+
+Once the recipient node starts receiving content from multiple peer nodes, it prunes down the long-latency peers and keeps the one to which the RTT is the shortest. Current proposals within the IPFS ecosystem are considering keeping the node with the highest throughput instead. It is not clear at this point which is the best approach.
+
+It is important to highlight that bitswap is a message-oriented protocol and _not_ a request-response protocol. That is, it sends messages including WANT lists and  is then waiting for the blocks in the WANT list to be delivered back, as and when the discovery protocol manages to find the blocks in the WANT list.
+
+
+This process of bitswap results in the following major problems:
+
+- **Many Roundtrips:** The most pressing issue is that requesting nodes have to make several roundtrips to walk the DAG, especially given that IPFS is a network run by untrusted nodes.
+- **Duplicate Data:** Bitswap is operating at the block level. In turn, this means that every block has to be requested from each peer that the node is connected to. This process results in duplicate traffic towards the data requestor, unnecessary load for the peers, and high overhead for the network.
```

Comment:
In the current release of Bitswap there is the concept of Sessions, used to keep track of peers that have some of the blocks that the Session is interested in:
- When IPFS asks Bitswap for the blocks corresponding to a set of CIDs, Bitswap
  - creates a new session
  - starts a discovery phase to find the peers with a subset of the blocks
- As peers are discovered they are remembered by the Session
- The Session sends subsequent requests only to those peers
- If there is a timeout, the Session tries to discover more peers through the DHT